### PR TITLE
Breaking changes time!~ (WithComponents -> AddComponents)

### PR DESCRIPTION
**Pull request by VelvetThePanda**
# Summary
A handful of people 
were complaining about naming conventions regarding .WithComponents, and now it's .AddComponents, to be a bit more explicit.

# Details
- Changed .WithComponents to .AddComponents to all applicable builders
- Removed the internal constructor of DiscordActionRowComponent (people complained about this too).
- Added ActionRow overloads to all applicable builders.

# Notes
Community bad x4

### CODE REVIEW

**Comment by VelvetThePanda** [comment]
Give me a second while I rebase this branch and more than likely end up nuking the entire tree by accident

**Comment by jeffreyladd** [comment]
This should be rejected as per the builder pattern, With should always be used over Add.  This was discussed in #695

**Comment by VelvetThePanda** [comment]
> This should be rejected as per the builder pattern, With should always be used over Add. This was discussed in #695

...Is there a problem with accumulative methods? I personally don't read through almost a hundred messages to try and figure out what's wrong specifically.

**Comment by VelvetThePanda** [comment]
I take it [this is the relevant message](https://github.com/DSharpPlus/DSharpPlus/pull/695#issuecomment-753539329), but as I've mentioned before, buttons behave in such a way that doesn't conform exactly to the...*ideals* you've projected in the past.

**Comment by VelvetThePanda** [comment]
The overload taking an ActionRowComponent would be the most fitting given what you've said in regards to what should and shouldn't be done with the builder, but forcing that one overload would be a bit clunky, imo.

**Comment by jeffreyladd** [comment]
Its all about the context.  You add to something that already exists.  Now since you are building either a Set of edits OR a new message entirely, its always with.

# Edit context
I am making x edits ##WITH x,y,z

# Create context
I am creating a new message with x files, y compontent, z embeds

**Comment by VelvetThePanda** [comment]
> Its all about the context. You add to something that already exists.
Yes. And technically, that's what it's doing. Internally.

**Comment by jeffreyladd** [comment]
Again u are building something new that edits something else.

**Comment by VelvetThePanda** [comment]
No???

**Comment by Lulalaby** [comment]
> 
> 
> Again u are building something new that edits something else.

Are you an maintainer? no.
Let emzi & the rest decide if its wrong.
Just stop it.

**Comment by DWaffles** [comment]
From a third party POV and my own opinion. `WithComponents` implies it will overwrite the previously given components, whereas `Add`... well... adds them (on a new line). 

I understand there's a pattern with the builder, and that's fine. It just seems that for a number of people this particular instance is a bit confusing.

**Comment by VelvetThePanda** [comment]
> From a third party POV and my own opinion. `WithComponents` implies it will overwrite the previously given components, whereas `Add`... well... adds them (on a new line).
> 
> I understand there's a pattern with the builder, and that's fine. It just seems that for a number of people this particular instance is a bit confusing.

Good thing .WithFiles isn't too popular either, because it does the same thing (adding, instead of overwriting)

**Comment by jeffreyladd** [comment]
> > Again u are building something new that edits something else.
> 
> Are you an maintainer? no.
> Let emzi & the rest decide if its wrong.
> Just stop it.

Neither are you.  I wasn't saying I have the final decision either, I stated why it should be rejected. 




> From a third party POV and my own opinion. `WithComponents` implies it will overwrite the previously given components, whereas `Add`... well... adds them (on a new line).
> 
> I understand there's a pattern with the builder, and that's fine. It just seems that for a number of people this particular instance is a bit confusing.

Builders can be confusing in the very beginning I agree, I also had the same confusion when I first started with it, but we should all understand what we are using.

**Comment by VelvetThePanda** [comment]
I still feel like the changes should remain because of how components work. You can read about them [here](https://discord.dev/interactions/message-components).

As I said, I deliberately structured it this way because components are *supposed to be accumulative*. If you don't like it, I'm sorry. I'm just trying to keep the lib up to date with cool features people would otherwise be repeating over and over in #support on the server like they did when slash commands first came out.

**Comment by jeffreyladd** [comment]
> I still feel like the changes should remain because of how components work. You can read about them [here](https://discord.dev/interactions/message-components).
> 
> As I said, I deliberately structured it this way because components are _supposed to be accumulative_. If you don't like it, I'm sorry. I'm just trying to keep the lib up to date with cool features people would otherwise be repeating over and over in #support on the server like they did when slash commands first came out.

I have read about them and also am implementing them in a separate lib.  But just to get them rushed out doesn't give justification to break patterns and standards.  That is just one of many reasons why libs go into disarray.

**Comment by DWaffles** [comment]
> Good thing .WithFiles isn't too popular either, because it does the same thing (adding, instead of overwriting)

Yeah, I'm not sure how this is with non-native English speakers, but the semantic difference between `Add` and `With` for the builder is... I don't want to say negligible but intellisense and documentation are enough to allow this convention to have documented exceptions for the sake of clarity.

One can rename `WithComponents` to `WithComponentsRow` I guess?

> Builders can be confusing in the very beginning I agree, I also had the same confusion when I first started with it, but we should all understand what we are using.

We should all understand it yes, and the documentation is fairly straight forward for the builder. It just seems given how components work both from a Discord and library standpoint, that this naming change is going to clear up confusions people currently have. Components are fairly distinct from the rest of Discord's features, I think it's reasonable that this is reflected in how people add components to messages.

**Comment by Lulalaby** [comment]
> We should all understand it yes, and the documentation is fairly straight forward for the builder. It just seems given how components work both from a Discord and library standpoint, that this naming change is going to clear up confusions people currently have. Components are fairly distinct from the rest of Discord's features, I think it's reasonable that this is reflected in how people add components to messages.

Exactly this

**Comment by Lulalaby** [comment]
> also am implementing them in a separate lib

Without wanting to be rude, then please go back to your lib.
Velvet implements choosed the actual way and that's it.
I guess you can see that this discussion is leading to nowhere.

**Comment by jeffreyladd** [comment]
> > also am implementing them in a separate lib
> 
> Without wanting to be rude, then please go back to your lib.
> Velvet choosed the actual way and that's it.
> I guess you can see that this discussion is leading to nowhere.

LOL the balls on you.  How u include something that is actually constructive to the conversation instead of trolling.

**Comment by DWaffles** [comment]
I think at this point all of us can continue discussing the minutia of this change with little to show for it. Unless any new discussion points get brought up, we should let this talking point rest.

**Comment by Lulalaby** [comment]
> 
> 
> > > also am implementing them in a separate lib
> > 
> > 
> > Without wanting to be rude, then please go back to your lib.
> > Velvet choosed the actual way and that's it.
> > I guess you can see that this discussion is leading to nowhere.
> 
> LOL the balls on you. How u include something that is actually constructive to the conversation instead of trolling.

The way how you stick to your opinion and answering yes when Velvet says no is the thing here.
I'm not trolling, I'm just at Velvets side.
I looked into it too and for me it's actually right how it's done rn.

**Comment by VelvetThePanda** [comment]
> [...] doesn't give justification to break patterns and standards.

I've said it once and I'll say it again. You focus *too* much about patterns. They're good, yes. But you can't have ideals *and* conform to an API *and* make users happy. I try to not diverge from what the lib does, but life ain't perfect. 

Regardless, I've had this discussion several times with you off-platform, and given that's being brought up *again* isn't indicative that it's going to be very productive this time either. 

If you have your standards and patterns you want to treasure like the holy grail, then sure, more power to you, but I'm not gonna keep repeating myself.

**Comment by Emzi0767** [comment]
Please limit the conversation on this PR to civilized discourse on the matter of the request itself.

**Comment by Emzi0767** [comment]
Regarding `.Add/.With`.

`.AddX` implies that a new X is going to be added to the builder. `.WithX` implies replacement.

**Comment by VelvetThePanda** [comment]
Gonna rebase this to remove the irrelevant commits (git, why are you like this x2), and it should be ready for review.

**Comment by DWaffles** [comment]
> 
> 
> Regarding `.Add/.With`.
> 
> `.AddX` implies that a new X is going to be added to the builder. `.WithX` implies replacement.

Ggiven that rows of components are added on multiple calls, I'm in favor of the change for semantic clarity and clearing the confusion currently in #support.

**Comment by jeffreyladd** [comment]
> I looked into it too and for me it's actually right how it's done rn.

this again as you said, is a subject of opinion.  

> If you have your standards and patterns you want to treasure like the holy grail, then sure, more power to you, but I'm not gonna keep repeating myself.

I don't hold onto them like a holy grail and implying that I do shows how little to what I have said to you off platform you actually listened.  As I told you, patterns and standards should be followed unless there is an excellent reason.  User convivence is not one of those reasons.

**Comment by VelvetThePanda** [comment]
Okay. Much better diff. cc @Emzi0767

**Comment by Emzi0767** [comment]
Squish squash your git skills are trash

**Review by Emzi0767 — APPROVED** [review]

File: DSharpPlus/Entities/Interaction/Components/DiscordActionRowComponent.cs
Comment (Emzi0767) [inline DSharpPlus/Entities/Interaction/Components/DiscordActionRowComponent.cs:51]: akshually this is mutable

Comment (VelvetThePanda) [reply DSharpPlus/Entities/Interaction/Components/DiscordActionRowComponent.cs:51]: Ack. Yeah, I was thinking about how this should be handled since someone may want to remove a component after the fact, and then they'd have to new up a new action row. Bleh. 

I forget who told me to make the components immutable.

**Comment by VelvetThePanda** [comment]
Anyway @Emzi0767 I guess I'll just try to figure out what to do in regards to working around the immutability myself. Changes have been reverted.

**Comment by Emzi0767** [comment]
do I merg?

**Comment by VelvetThePanda** [comment]
Indeed

**Pull request merged**
